### PR TITLE
feat(generator): improve method comments

### LINKED ISSUE #11674: Make RPC parameter links more useful
The links to `*Request` and response types is not very helpful.  Maybe they could say "this is a C++ class generated by Protobuf from the foo.bar.service.v1.FooRequest" proto message.

There are similar problems for the return links, and for long-running operations and paginations.

  image

### PR DESCRIPTION
**Pull request by coryan**
Add more details about the returned types for different method
"category" (paginated, long-running, bidir, etc.). Include links to the
proto message and to the mapping rules. Explain the role of `StatusOr`
and `future` in these RPCs.

Also include a longer description of the `request` parameter, including
links to the Protobuf mapping rules and the original proto message.

I disabled formatting around the comments for each method. It was too
hard to create stable reference-style links, and I think we don't want
to mess with the newlines introduced in the proto comments.

Fixes #11674

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11677)

### CODE REVIEW

File: ci/cloudbuild/builds/generate-libraries.sh [hunk 1]
```diff
@@ -71,6 +71,8 @@ if [ -z "${GENERATE_GOLDEN_ONLY}" ]; then
   io::log_h2 "Formatting generated code"
   git ls-files -z -- '*.h' '*.cc' '*.proto' |
     xargs -P "$(nproc)" -n 1 -0 clang-format -i
+  git ls-files -z -- '*.h' |
+    xargs -r -P "$(nproc)" -n 50 -0 sed -i 's/[[:blank:]]\+$//'
```
Comment (devbww) [inline ci/cloudbuild/builds/generate-libraries.sh:75]: Should a similar thing be done to generated golden headers in the "GENERATE_GOLDEN_ONLY" case?

File: generator/internal/descriptor_utils.cc [hunk 2]
```diff
@@ -527,10 +527,16 @@ std::string FormatMethodCommentsMethodSignature(
...
     google::protobuf::MethodDescriptor const& method) {
-  google::protobuf::Descriptor const* input_type = method.input_type();
-  auto parameter_comment = absl::StrFormat("  /// @param %s %s\n", "request",
-                                           FormatDoxygenLink(*input_type));
-  return FormatMethodComments(method, std::move(parameter_comment));
+  auto constexpr kRequestParam =
+      R"""(  /// @param request Unary RPCs, such as the one wrapped by this
+  ///     function, receive a single `request` proto message which includes all
+  ///     the inputs for the RPC. In this case, the proto message is a
+  ///     [%s].
```
Comment (devbww) [inline generator/internal/descriptor_utils.cc:534]: Should the type name be back-tick quoted?

File: generator/internal/format_method_comments.cc [hunk 3]
```diff
@@ -37,6 +37,81 @@ ProtoDefinitionLocation Location(google::protobuf::Descriptor const& d) {
...
+  }
+  if (IsBidirStreaming(method)) {
+    return absl::StrFormat(
+        R"""(  /// @return An object representing the bidirectional streaming
+  ///     RPC. Applications can send multiple request messages and receive
+  ///     multiple response messages through this API. Bidirectional streaming
+  ///     RPCs can impose restrictions on the sequence of request and response
+  ///     messages. Please consult the service documentation for details.
+  ///     The request message type ([%s]) and response messages
+  ///     ([%s]) are mapped to C++ classes using the
```
Comment (devbww) [inline generator/internal/format_method_comments.cc:63]: Should the type names be back-tick quoted?

File: generator/internal/format_method_comments.cc [hunk 4]
```diff
@@ -37,6 +37,81 @@ ProtoDefinitionLocation Location(google::protobuf::Descriptor const& d) {
...
+  ///     to iterate of the results. See the documentation of this type for
+  ///     details. In brief, this class has `begin()` and `end()` member
+  ///     functions returning a iterator class meeting the
+  ///     [input iterator requirements]. The value type for this iterator is a
+  ///     [`StatusOr`] as the iteration may fail even after some values are
+  ///     retrieved successfully, for example, if there is a network disconnect.
+  ///     An empty set of results does not indicate an error, it indicates
+  ///     that there are no resources meeting the request criteria.
+  ///     On a successful iteration the `StatusOr<T>` contains elements of type
+  ///     [%s], or rather,
```
Comment (devbww) [inline generator/internal/format_method_comments.cc:95]: Should the type name be back-tick quoted?

File: generator/internal/format_method_comments.cc [hunk 5]
```diff
@@ -37,6 +37,81 @@ ProtoDefinitionLocation Location(google::protobuf::Descriptor const& d) {
...
+  ///     the C++ class generated by Protobuf from that type. Please consult the
+  ///     Protobuf documentation for details on the [Protobuf mapping rules].
+)""",
+        info->second->full_name());
+  }
+  if (IsResponseTypeEmpty(method)) {
+    return R"""(  /// @return a [`Status`] object. If the request failed the
```
Comment (devbww) [inline generator/internal/format_method_comments.cc:102]: s/failed the/failed, the/

File: generator/internal/format_method_comments.cc [hunk 6]
```diff
@@ -37,6 +37,81 @@ ProtoDefinitionLocation Location(google::protobuf::Descriptor const& d) {
...
+    google::protobuf::MethodDescriptor const& method) {
+  if (IsLongrunningOperation(method)) {
+    return R"""(  /// @return A [`future`] that becomes satisfied when the LRO
+  ///     ([Long Running Operation]) completes or the polling policy in effect
+  ///     for this call is exhausted. The future is satisfied with an error if
+  ///     the LRO completes with an error or the polling policy is exhausted.
+  ///     In this case the [`StatusOr`] returned by the future contains the
+  ///     error. If the LRO completes successfully the value of the future
+  ///     contains the LRO's result. For this RPC the result is a
+  ///     [$longrunning_deduced_response_message_type$] proto message.
```
[the remark this answers is not in the document]
Comment (coryan) [reply generator/internal/format_method_comments.cc:50]: Done.

File: generator/internal/format_method_comments.cc (hunk 3 above)
Comment (coryan) [reply generator/internal/format_method_comments.cc:63]: I think so.  I even implemented the fix, but it broke other links extracted from the proto comments.  I created #11685 to track this.

File: generator/internal/format_method_comments.cc (hunk 5 above)
Comment (coryan) [reply generator/internal/format_method_comments.cc]: Fixed

**Review by devbww — APPROVED** [review]

File: ci/cloudbuild/builds/generate-libraries.sh [hunk 7]
```diff
@@ -71,14 +71,18 @@ if [ -z "${GENERATE_GOLDEN_ONLY}" ]; then
...
-    'generator/integration_tests/golden/v1/*.h' \
-    'generator/integration_tests/golden/v1/internal/*.cc' \
-    'generator/integration_tests/golden/v1/*.cc' \
+  git ls-files -z -- 'generator/integration_tests/golden/**/*.h' \
+    'generator/integration_tests/golden/v1/**/*.cc' \
     'generator/integration_tests/*.proto' |
     xargs -P "$(nproc)" -n 1 -0 clang-format -i
+  git ls-files -z -- 'generator/integration_tests/golden/**/*.h' \
+    'generator/integration_tests/golden/v1/**/*.cc' \
+    'generator/integration_tests/*.proto' |
```
[the remark this answers is not in the document]
Comment (coryan) [reply ci/cloudbuild/builds/generate-libraries.sh:84]: Fixed.

**Pull request merged**